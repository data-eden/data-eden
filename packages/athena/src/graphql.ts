import type { TypedDocumentNode } from '@graphql-typed-document-node/core';
import type { DefaultVariables } from './types.js';

export function graphql<
  Data extends object = object,
  Variables extends DefaultVariables = DefaultVariables
>(
  literals: TemplateStringsArray,
  ...args: any[]
): TypedDocumentNode<Data, Variables> {
  throw new Error(
    'This function should never actually be evaluated. Please use the babel plugin.'
  );
}
