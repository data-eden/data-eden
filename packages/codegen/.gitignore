__graphql-project
.rollup.cache
