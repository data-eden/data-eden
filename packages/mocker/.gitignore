__tests__/__generated