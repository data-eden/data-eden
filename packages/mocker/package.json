{
  "name": "@data-eden/mocker",
  "version": "0.16.0",
  "description": "",
  "homepage": "https://github.com/data-eden/data-eden#readme",
  "bugs": {
    "url": "https://github.com/data-eden/data-eden/issues"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/data-eden/data-eden.git"
  },
  "license": "ISC",
  "author": "Gabriel Csapo <gabecsapo@gmail.com>",
  "type": "module",
  "exports": {
    ".": {
      "types": "./dist/types/index.d.ts",
      "import": "./dist/data-eden-mocker.js",
      "require": "./dist/data-eden-mocker.umd.cjs"
    }
  },
  "main": "./dist/data-eden-mocker.umd.cjs",
  "module": "./dist/data-eden-mocker.js",
  "types": "./dist/types/index.d.ts",
  "scripts": {
    "build": "vite build && tsc -p tsconfig.build.json",
    "dev": "tsc --watch",
    "test": "vitest run"
  },
  "dependencies": {
    "@faker-js/faker": "8.0.2",
    "graphql": "^16.6.0"
  },
  "devDependencies": {
    "@babel/preset-typescript": "^7.22.5",
    "@data-eden/codegen": "^0.16.0",
    "vitest": "^0.32.0"
  },
  "volta": {
    "extends": "../../package.json"
  },
  "publishConfig": {
    "access": "public"
  }
}
