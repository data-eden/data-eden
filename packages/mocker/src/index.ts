import type { TypedDocumentNode } from '@graphql-typed-document-node/core';
import type {
  GraphQLSchema,
  GraphQLNamedType,
  SelectionNode,
  GraphQLType,
  GraphQLEnumValue,
} from 'graphql';
import {
  buildSchema,
  GraphQLObjectType,
  GraphQLList,
  GraphQLEnumType,
  GraphQLNonNull,
  isNonNullType,
  GraphQLUnionType,
  Kind,
} from 'graphql';
import type { Faker } from '@faker-js/faker';
import { faker as defaultFaker } from '@faker-js/faker';

import { inlineAllFragments } from './inline-all-fragments.js';

type JSONPrimitive = string | number | boolean | null;
type JSONArray = JSONValue[];
type JSONObject = { [key in string]: JSONValue | undefined };
type JSONValue = JSONPrimitive | JSONArray | JSONObject;

const defaultTypeGenerators: TypeGenerators = {
  String(faker: Faker) {
    return faker.word.words();
  },
  Int(faker: Faker) {
    return faker.number.int();
  },
  Long(faker: Faker) {
    let _min = BigInt(0n);
    let _max = BigInt(9223372036854775807n);

    return (_min +
      BigInt(
        Math.floor(Math.random() * Number(_max - _min + 1n))
      )) as unknown as JSONValue;
  },
  Float(faker: Faker) {
    const randomNum = faker.datatype.number({
      min: 0,
      max: 92233720368547,
      precision: 0.01,
    });

    return parseFloat(randomNum.toFixed(2));
  },
};

type DeepPartial<T> = T extends any[]
  ? DeepPartialArray<T[number]>
  : T extends object
  ? DeepPartialObject<T>
  : T;

type DeepPartialArray<T> = Array<DeepPartial<T>>;

type DeepPartialObject<T> = {
  [P in keyof T]?: DeepPartial<T[P]>;
};

type FieldGenerators = {
  [className: string]: {
    [key: string]: () => JSONValue;
  };
};

type TypeGenerators = {
  [key: string]: (faker: Faker) => JSONValue;
};

interface MockerOptions {
  schema: string;
  faker?: Faker;
  typeGenerators?: TypeGenerators;
  fieldGenerators?: FieldGenerators;
}

type FieldWithSelectionSet = SelectionNode & {
  selectionSet: {
    selections: readonly SelectionNode[];
  };
};

function assertHasSelectionSet(
  field: SelectionNode | undefined
): asserts field is FieldWithSelectionSet {
  if (field && field.kind !== Kind.FRAGMENT_SPREAD && !field.selectionSet) {
    const fieldName =
      (field.kind === Kind.INLINE_FRAGMENT
        ? field?.typeCondition?.name.value
        : field.name.value) ?? '"Can not resolve the name"';
    throw new Error(`Field ${fieldName} does not have a selection set.`);
  }
}

type QueryOrMutationWrapped<T> = {
  __meta__: {
    $DEBUG: {
      ast: TypedDocumentNode<T>;
    };
  };
};

export class Mocker {
  private faker: Faker;
  private fieldGenerators: FieldGenerators;
  private typeGenerators: TypeGenerators;
  private schema: GraphQLSchema;

  constructor({
    schema,
    faker,
    typeGenerators,
    fieldGenerators,
  }: MockerOptions) {
    if (!schema) {
      throw new Error('Schema is required');
    }

    this.faker = faker ?? defaultFaker;
    this.fieldGenerators = fieldGenerators || {};
    this.typeGenerators = {
      ...defaultTypeGenerators,
      ...(typeGenerators || {}),
    };
    this.schema = buildSchema(schema);

    // we need to set the seed on the faker instance
    this.faker.seed(2345678);
  }

  mock<Document extends JSONObject = JSONObject>(
    documentNode:
      | TypedDocumentNode<Document>
      | QueryOrMutationWrapped<Document>,
    mockData: DeepPartial<Document>
  ): Document {
    const definitionNode =
      '__meta__' in documentNode
        ? documentNode.__meta__.$DEBUG.ast.definitions[0]
        : documentNode.definitions[0];

    const definition = inlineAllFragments(
      definitionNode,
      '__meta__' in documentNode
        ? documentNode.__meta__.$DEBUG.ast
        : documentNode
    );

    let typeName: string;
    let selections: readonly SelectionNode[];

    if (definition.kind === 'FragmentDefinition') {
      typeName = definition.typeCondition.name.value;
      selections = definition.selectionSet.selections;
    } else if (
      definition.kind === 'OperationDefinition' &&
      definition.operation === 'query'
    ) {
      typeName = 'Query';
      selections = definition.selectionSet.selections;
    } else if (
      definition.kind === 'OperationDefinition' &&
      definition.operation === 'mutation'
    ) {
      typeName = 'Mutation';
      selections = definition.selectionSet.selections;
    } else {
      throw new Error(`Unsupported definition kind "${definition.kind}"`);
    }

    const type = this.schema.getType(typeName) as GraphQLNamedType;

    if (!type) {
      throw new Error(`Type ${typeName} not found in the schema`);
    }

    if (type instanceof GraphQLUnionType) {
      return this.generateMockDataForUnionFields<Document>(
        definition as unknown as SelectionNode,
        type,
        mockData
      );
    }

    return this.generateMockDataForFields<Document>(selections, type, mockData);
  }

  private generateMockDataForUnionFields<
    Data extends JSONValue | undefined | null = JSONValue | undefined | null
  >(
    field: SelectionNode,
    fieldType: GraphQLUnionType,
    mockData: JSONObject | undefined | null
  ): Data {
    const possibleTypes = fieldType.getTypes();

    // Check if mockData is provided for the union field
    if (mockData) {
      const mockDataUnionType = mockData.__typename;
      const matchingType = possibleTypes.find(
        (type) => type.name === mockDataUnionType
      );
      if (matchingType) {
        assertHasSelectionSet(field);
        const selectionSetForType = field.selectionSet.selections.find(
          (selection) => {
            // we need to return the selection set that matches this type
            return (
              selection.kind === Kind.INLINE_FRAGMENT &&
              selection.typeCondition &&
              selection.typeCondition.name.value === matchingType.name
            );
          }
        );

        assertHasSelectionSet(selectionSetForType);

        return this.generateMockDataForFields(
          selectionSetForType.selectionSet.selections,
          matchingType,
          mockData
        );
      }
    }

    assertHasSelectionSet(field);
    const randomIndex = Math.floor(Math.random() * possibleTypes.length);
    const randomType = possibleTypes[randomIndex];
    const selectionSetForType = field.selectionSet.selections.find(
      (selection) => {
        // we need to return the selection set that matches this type
        return (
          selection.kind === Kind.INLINE_FRAGMENT &&
          selection.typeCondition &&
          selection.typeCondition.name.value === randomType.name
        );
      }
    );
    assertHasSelectionSet(selectionSetForType);
    return this.generateMockDataForFields(
      selectionSetForType.selectionSet.selections,
      randomType,
      mockData
    );
  }

  private generateMockDataForFields<
    Data extends JSONValue | undefined | null = JSONValue | undefined | null
  >(
    fields: readonly SelectionNode[],
    type: GraphQLNamedType,
    mockData: JSONObject | undefined | null
  ): Data {
    let result: JSONObject = {};

    fields.forEach((field) => {
      if (field.kind === 'InlineFragment') {
        // Handle inline fragments
        if (!field.typeCondition) {
          throw new Error(
            'Could not handle a case where inline fragment typeCondition was undefined'
          );
        }

        const fragmentType = field.typeCondition.name.value;
        result = {
          ...result,
          ...this.generateMockDataForFields(
            field.selectionSet.selections,
            this.schema.getType(fragmentType) as GraphQLNamedType,
            mockData
          ),
        };
      } else if (field.kind === 'FragmentSpread') {
        throw new Error(
          'Fragment spreads are not supported, please resolve all fragments spreads before trying to mock.'
        );
      } else {
        const fieldName = field.name.value;

        if (fieldName === '__typename') {
          result[fieldName] = type.name;
        } else {
          if (
            !type ||
            !('getFields' in type) ||
            !type.getFields ||
            !type.getFields()[fieldName]
          ) {
            throw new Error(
              `Field ${fieldName} does not exist in the schema for type ${type?.name}`
            );
          }

          let fieldType = type.getFields()[fieldName].type;

          // If it's a NonNull type, unwrap it
          if (fieldType instanceof GraphQLNonNull) {
            fieldType = fieldType.ofType;
          }

          // Handle list types (arrays)
          if (fieldType instanceof GraphQLList) {
            let elementType: GraphQLType = fieldType.ofType;

            // If it's a NonNull type, unwrap it again to get the actual element type
            if (elementType instanceof GraphQLNonNull) {
              elementType = elementType.ofType;
            }

            // If no mock data is provided for the list, create an empty array
            if (!mockData || !mockData[fieldName]) {
              result[fieldName] = [];
            } else {
              // Check if the element is an object type, and if so, recursively generate the mock data
              if (elementType instanceof GraphQLObjectType) {
                result[fieldName] = (mockData[fieldName] as JSONArray).map(
                  (item) => {
                    assertHasSelectionSet(field);

                    return this.generateMockDataForFields(
                      field.selectionSet.selections,
                      elementType as GraphQLObjectType,
                      item as JSONObject
                    );
                  }
                ) as JSONValue;
              } else {
                // Handle scalar type arrays (e.g., [Int], [String])
                result[fieldName] = (mockData[fieldName] as JSONArray).map(
                  (item) =>
                    this.generateMockDataForType(
                      type.name,
                      fieldName,
                      elementType,
                      item
                    )
                );
              }
            }
          } else if (fieldType instanceof GraphQLObjectType) {
            // Recursively handle object types
            assertHasSelectionSet(field);

            result[fieldName] = this.generateMockDataForFields(
              field.selectionSet.selections,
              fieldType,
              mockData && (mockData[fieldName] as JSONObject)
            );
          } else if (fieldType instanceof GraphQLEnumType) {
            // Handle enum types
            const enumValues = fieldType.getValues();
            result[fieldName] = this.generateMockDataForEnum(
              fieldType.name,
              enumValues,
              mockData && mockData[fieldName]
            );
          } else if (fieldType instanceof GraphQLUnionType) {
            // Handle union types;
            result[fieldName] = this.generateMockDataForUnionFields(
              field,
              fieldType,
              mockData && (mockData[fieldName] as JSONObject)
            );
          } else {
            // Generate mock data for scalar fields
            result[fieldName] = this.generateMockDataForType(
              type.name,
              fieldName,
              fieldType,
              mockData && (mockData[fieldName] as JSONObject)
            );
          }
        }
      }
    });

    return result as Data;
  }

  private generateMockDataForEnum(
    fieldName: string,
    enumValues: readonly GraphQLEnumValue[],
    providedValue: JSONValue | undefined
  ): string {
    const values = enumValues.map((enumValue) => enumValue.name);
    if (providedValue) {
      if (typeof providedValue !== 'string') {
        throw new Error(
          `${JSON.stringify(
            providedValue
          )} is not a valid ENUM value as it can only be a string`
        );
      }

      if (values.indexOf(providedValue) > -1) {
        return providedValue;
      } else {
        throw new Error(
          `Trying to mock ${fieldName} with enum value ${providedValue} does not match known enum values "[${values.join(
            ','
          )}]"`
        );
      }
    }
    const randomIndex = Math.floor(Math.random() * values.length);
    return values[randomIndex];
  }

  private generateMockDataForType(
    className: string,
    fieldName: string,
    type: GraphQLType | GraphQLNonNull<GraphQLType>,
    inputMockData: JSONValue | undefined
  ): JSONValue {
    // Check if this is a NonNull type
    if (isNonNullType(type) && type.ofType) {
      return this.generateMockDataForType(
        className,
        fieldName,
        type.ofType,
        inputMockData
      );
    }

    // Use input mock data if provided
    if (inputMockData !== undefined) {
      return inputMockData;
    }

    if (
      this.fieldGenerators[className] &&
      this.fieldGenerators[className][fieldName]
    ) {
      return this.fieldGenerators[className][fieldName]();
    }

    if (!isNonNullType(type) && this.typeGenerators[type.name]) {
      return this.typeGenerators[type.name](this.faker);
    } else {
      if (isNonNullType(type)) {
        throw new Error(
          `Value generation could not be done for type ${type.ofType.toString()}. Please provide an override to typeGenerators.`
        );
      } else {
        throw new Error(
          `Value generation could not be done for type ${type.name} in ${className}. Please provide an override to typeGenerators.`
        );
      }
    }
  }
}
