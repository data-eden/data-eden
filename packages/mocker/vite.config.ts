import { defineConfig } from 'vite';
import { resolve } from 'path';

export default defineConfig({
  build: {
    sourcemap: true,
    lib: {
      entry: resolve(__dirname, 'src/index.ts'),
      name: '@data-eden/mocker',
      fileName: 'data-eden-mocker',
    },
  },
});
