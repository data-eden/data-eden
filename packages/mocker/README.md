# @data-eden/mocker

## Installation

```sh
yarn add @data-eden/codegen @data-eden/mocker
```

## Usage

> Utilizing the `gql` preprocesing from `@data-eden/codegen` passing that object to the mocker will be all you need along with the schema to start to generate mocks. By default the mocker mocks out required fields only, passing in the fields will override the default mocking strategy as well as mock out optionals if present in the data passed in.

```js
import { gql } from '@data-eden/codegen/gql';

const carTwoFragment = gql`
  fragment carTwo on Car {
    id
    make
  }
`;

const mocker = new Mocker({
  schema,
});

const result = mocker.mock(carTwoFragment, { id: 1234 });
/**
{
  "id": 1234,
  "make": "whose nor",
}
*/
```
