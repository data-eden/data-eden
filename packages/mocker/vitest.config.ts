import { defineConfig } from 'vitest/config';
import { resolve } from 'node:path';

import { rollupPlugin } from '@data-eden/codegen';

export default defineConfig({
  plugins: [
    // eslint-disable-next-line @typescript-eslint/no-unsafe-call
    rollupPlugin({
      documents: ['__tests__/**/*.ts'],
      baseDir: __dirname,
      schemaPath:
        '../../internal-packages/react-graphql-test-app/src/graphql/schema.graphql',
      extension: '.graphql.ts',
      disableSchemaTypesGeneration: false,
      production: false,
    }),
  ],
  test: {
    testTimeout: 10_000,
  },
  resolve: {
    alias: {
      '@data-eden/mocker': resolve(__dirname, './src'),
    },
  },
});
