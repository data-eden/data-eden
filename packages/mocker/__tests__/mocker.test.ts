import { describe, test, expect } from 'vitest';
import { readFileSync } from 'fs';
import { resolve } from 'path';

import { gql } from '@data-eden/codegen/gql';

import { Mocker } from '@data-eden/mocker';
import {
  CarOneFragment,
  CarOneHalfFragment,
  CarTwoFragment,
  CarTwoHalfFragment,
  CarThreeFragment,
  CarFourFragment,
  OwnerFragment,
  CreateOnePetMutation,
  PersonOneFragment,
  PetOneFragment,
  PetThreeFragment,
  PetTwoFragment,
  OwnerOneFragment,
} from './__generated/mocker.test.graphql';
import { Breed } from '../../../internal-packages/react-graphql-test-app/src/graphql/schema.graphql.js';

const schema = readFileSync(
  resolve(
    __dirname,
    '..',
    '..',
    '..',
    'internal-packages/react-graphql-test-app/src/graphql/schema.graphql'
  ),
  'utf8'
);

describe('mocker', () => {
  describe('fragments', () => {
    test('fragment with default data', async () => {
      const carOneFragment = gql<CarOneFragment>`
        fragment carOne on Car {
          id
          make
        }
      `;

      const mocker = new Mocker({ schema });
      const result = mocker.mock(carOneFragment, { id: '1234' });

      expect(result).toMatchSnapshot();
    });

    test('fragment with aliased data', async () => {
      const carOneHalfFragment = gql<CarOneHalfFragment>`
        fragment carOneHalf on Car {
          id
          carMake: make
        }
      `;

      const mocker = new Mocker({ schema });
      const result = mocker.mock(carOneHalfFragment, {
        id: '1234',
        carMake: 'accura tsx',
      });

      expect(result).toMatchSnapshot();
    });

    test('fragment with default data and fieldGenerator', async () => {
      const carTwoFragment = gql<CarTwoFragment>`
        fragment carTwo on Car {
          id
          make
        }
      `;

      const mocker = new Mocker({
        schema,
        fieldGenerators: {
          Car: {
            make() {
              return 'Acura';
            },
          },
        },
      });
      const result = mocker.mock(carTwoFragment, {
        id: '1234',
      });

      expect(result).toMatchSnapshot();
    });

    test('fragment with default data and fieldGenerator with validation', async () => {
      const carTwoHalfFragment = gql<CarTwoHalfFragment>`
        fragment carTwoHalf on Car {
          id
          make
        }
      `;

      const mocker = new Mocker({
        schema,
        fieldGenerators: {
          Car: {
            make(potentialValue) {
              const allowedMakes = ['Accura', 'Ford', 'GMC'];
              if (
                potentialValue &&
                typeof potentialValue === 'string' &&
                allowedMakes.indexOf(potentialValue) === -1
              ) {
                throw new Error(
                  `"${potentialValue}" is not a valid make (${allowedMakes.join(
                    ','
                  )})`
                );
              }
              return potentialValue ?? 'Acura';
            },
          },
        },
      });
      expect(() => {
        mocker.mock(carTwoHalfFragment, {
          id: '1234',
          make: 'Tesla',
        });
      }).toThrowErrorMatchingInlineSnapshot(
        '"\\"Tesla\\" is not a valid make (Accura,Ford,GMC)"'
      );
    });

    test('fragment with enum values', async () => {
      const petOneFragment = gql<PetOneFragment>`
        fragment petOne on Pet {
          id
          breed
        }
      `;

      const mocker = new Mocker({
        schema,
      });
      const result = mocker.mock(petOneFragment, { id: 1234 });

      expect(Object.keys(result)).toMatchInlineSnapshot(`
        [
          "id",
          "breed",
        ]
      `);
      expect(
        result.breed &&
          typeof result.breed === 'string' &&
          [
            'SHEPARD',
            'BULLDOG',
            'POODLE',
            'GERMAN_SHEPHERD',
            'LABRADOR_RETRIEVER',
            'GOLDEN_RETRIEVER',
          ].indexOf(result.breed) > -1
      ).toBeTruthy();
    });

    test('fragment with union values (random)', async () => {
      const carThreeFragment = gql<CarThreeFragment>`
        fragment carThree on Car {
          __typename
          id
          make
          owner {
            ... on Person {
              __typename
              id
              name
              car {
                __typename
                id
                make
              }
            }
            ... on Company {
              __typename
              id
              name
            }
          }
        }
      `;

      const mocker = new Mocker({
        schema,
        typeGenerators: {
          ID() {
            return '1234';
          },
        },
      });
      const result = mocker.mock(carThreeFragment, { id: '1234' });

      // we know that this is going to be a random result because we are picking a random resolution everytime
      // so the expects below are handling both cases for whatever union we mock out

      // top level object should be a car
      expect(Object.keys(result)).toEqual([
        '__typename',
        'id',
        'make',
        'owner',
      ]);
      // expect owner to be a specific type not all of the types as only one union can be projected
      expect(Object.keys(result.owner)).toEqual(
        result.owner.__typename === 'Company'
          ? ['__typename', 'id', 'name']
          : ['__typename', 'id', 'name', 'car']
      );
    });

    test('fragment with union values (with match)', async () => {
      const carFourFragment = gql<CarFourFragment>`
        fragment carFour on Car {
          __typename
          id
          make
          owner {
            ... on Person {
              __typename
              id
              name
              car {
                __typename
                id
                make
              }
            }
            ... on Company {
              __typename
              id
              name
            }
          }
        }
      `;

      const mocker = new Mocker({
        schema,
        typeGenerators: {
          ID() {
            return '123';
          },
        },
      });

      const result = mocker.mock(carFourFragment, {
        id: '1234',
        owner: {
          __typename: 'Person',
          id: '123',
          name: 'Bob',
        },
      });

      expect(result).toEqual({
        __typename: 'Car',
        id: '1234',
        make: 'whose nor',
        owner: {
          __typename: 'Person',
          id: '123',
          name: 'Bob',
          car: {
            __typename: 'Car',
            id: '123',
            make: 'imperfect offensively thwack',
          },
        },
      });
    });

    test('should be able to mock union type fragment', async () => {
      const ownerFragment = gql<OwnerFragment>`
        fragment owner on Owner {
          ... on Person {
            __typename
            id
            name
            car {
              __typename
              id
              make
            }
          }
          ... on Company {
            __typename
            id
            name
          }
        }
      `;

      const mocker = new Mocker({
        schema,
        typeGenerators: {
          ID: () => {
            return '123';
          },
        },
      });
      const result = mocker.mock(ownerFragment, {
        __typename: 'Person',
        id: '123',
        name: 'Bob',
        car: {
          make: 'Accura',
        },
      });

      expect(result).toEqual({
        __typename: 'Person',
        id: '123',
        name: 'Bob',
        car: {
          __typename: 'Car',
          id: '123',
          make: 'Accura',
        },
      });
    });

    test('should be able to mock union type fragment (from other references)', async () => {
      const personNestedFragment = gql`
        fragment personNested on Person {
          __typename
          id
          name
          car {
            __typename
            id
            make
          }
        }
      `;
      const companyNestedFragment = gql`
        fragment companyNested on Company {
          __typename
          id
          name
        }
      `;
      const ownerOneFragment = gql<OwnerOneFragment>`
        fragment ownerOne on Owner {
          ${personNestedFragment}
          ${companyNestedFragment}
        }
      `;

      const mocker = new Mocker({
        schema,
        typeGenerators: {
          ID: () => {
            return '123';
          },
        },
      });
      const result = mocker.mock(ownerOneFragment, {
        __typename: 'Person',
        id: '123',
        name: 'Bob',
        car: {
          make: 'Accura',
        },
      });

      expect(result).toEqual({
        __typename: 'Person',
        id: '123',
        name: 'Bob',
        car: {
          __typename: 'Car',
          id: '123',
          make: 'Accura',
        },
      });
    });

    test('fragment with enum values (with provided enum value)', async () => {
      const petTwoFragment = gql<PetTwoFragment>`
        fragment petTwo on Pet {
          id
          breed
        }
      `;

      const mocker = new Mocker({
        schema,
      });
      const result = mocker.mock(petTwoFragment, {
        id: '1234',
        breed: Breed.Shepard,
      });

      expect(result).toMatchSnapshot();
    });

    test("fragment with enum values (with provided enum value that doesn't match, this should throw)", async () => {
      const petThreeFragment = gql<PetThreeFragment>`
        fragment petThree on Pet {
          id
          breed
        }
      `;

      const mocker = new Mocker({
        schema,
      });

      expect(() =>
        mocker.mock(petThreeFragment, { id: '1234', breed: 'SHARK' as any })
      ).toThrowErrorMatchingSnapshot();
    });

    test('fragment with default data and fieldGenerator with array of values', async () => {
      const personOneFragment = gql<PersonOneFragment>`
        fragment personOne on Person {
          id
          name
          car {
            id
          }
          pets {
            id
            name
          }
        }
      `;

      const mocker = new Mocker({
        schema,
        typeGenerators: {
          ID() {
            return '1234557';
          },
        },
      });
      const result = mocker.mock(personOneFragment, {
        id: '1234',
        pets: [{}, { id: '99999', name: 'Bob' }],
      });

      expect(result).toMatchSnapshot();
    });
  });

  describe('queries', () => {
    test('should work with basic query (no overrides)', async () => {
      const carOneQuery = gql<CarOneFragment>`
        query carOne {
          car(id: 124) {
            id
            make
            model
          }
        }
      `;
      const mocker = new Mocker({
        schema,
        typeGenerators: {
          ID() {
            return '1234557';
          },
        },
        fieldGenerators: {
          Car: {
            make() {
              return 'Italian Car';
            },
          },
        },
      });

      const result = mocker.mock(carOneQuery, { id: '1234' });

      expect(result).toMatchSnapshot();
    });
  });

  describe('mutations', () => {
    test('should work with basic mutation (no overrides)', async () => {
      const createOnePetMutation = gql<CreateOnePetMutation>`
        mutation createOnePet {
          createPet(input: { name: "Bob", personId: 1234 }) {
            id
            name
          }
        }
      `;

      const mocker = new Mocker({
        schema,
        typeGenerators: {
          ID() {
            return '1234557';
          },
        },
      });

      const result = mocker.mock(createOnePetMutation, {
        createPet: { id: '1234' },
      });

      expect(result).toMatchSnapshot();
    });
  });
});
