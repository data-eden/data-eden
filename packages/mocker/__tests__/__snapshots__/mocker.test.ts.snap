// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`mocker > fragments > fragment with default data 1`] = `
{
  "id": 1234,
  "make": "whose nor",
}
`;

exports[`mocker > fragments > fragment with default data and fieldGenerator 1`] = `
{
  "id": 1234,
  "make": "Acura",
}
`;

exports[`mocker > fragments > fragment with default data and fieldGenerator with array of values 1`] = `
{
  "car": {
    "id": 1234,
  },
  "id": 1234,
  "name": "whose nor",
  "pets": [
    {
      "id": 1234557,
      "name": "imperfect offensively thwack",
    },
    {
      "id": 99999,
      "name": "Bob",
    },
  ],
}
`;

exports[`mocker > fragments > fragment with enum values (with provided enum value that doesn't match, this should throw) 1`] = `[Error: Trying to mock Breed with enum value SHARK does not match known enum values "[SHEPARD,BULLDOG,POODLE,GERMAN_SHEPHERD,LABRADOR_RETRIEVER,GOLDEN_RETRIEVER]"]`;

exports[`mocker > fragments > fragment with enum values (with provided enum value) 1`] = `
{
  "breed": "SHEPARD",
  "id": 1234,
}
`;

exports[`mocker > fragments > fragment with union values 1`] = `
{
  "id": 1234,
  "make": "whose nor",
  "owner": {
    "id": 1234,
    "name": "tie",
  },
}
`;

exports[`mocker > mutations > should work with basic mutation (no overrides) 1`] = `
{
  "createPet": {
    "__typename": "Pet",
    "id": 1234,
    "name": "whose nor",
  },
}
`;

exports[`mocker > queries > should work with basic query (no overrides) 1`] = `
{
  "car": {
    "__typename": "Car",
    "id": 1234,
    "make": "Italian Car",
    "model": "whose nor",
  },
}
`;
